test/gcs/helper_container_test.go

```diff
@@ -0,0 +1,290 @@
+		t.Fatalf("got exit code %d", e)
+	}
+}
+
+func waitContainerRaw(c *hcsv2.Container, p hcsv2.Process) (int, prot.NotificationType) {
+	exch, dch := p.Wait()
+	defer close(dch)
+	r := <-exch
+	dch <- true
+	n := c.Wait()
+
+	return r, n
+}
+
+func execProcess(ctx context.Context, t testing.TB, c *hcsv2.Container, p *oci.Process, con stdio.ConnectionSettings) hcsv2.Process {
+	pid, err := c.ExecProcess(ctx, p, con)
+	if err != nil {
+		t.Helper()
+		t.Fatalf("could not exec process: %v", err)
+	}
+
+	return getProcess(ctx, t, c, uint32(pid))
+}
+
+func getProcess(_ context.Context, t testing.TB, c *hcsv2.Container, pid uint32) hcsv2.Process {
+	p, err := c.GetProcess(pid)
+	if err != nil {
+		t.Helper()
+		t.Fatalf("could not get process %d: %v", pid, err)
+	}
+
+	return p
+}
+
+func killContainer(ctx context.Context, t testing.TB, c *hcsv2.Container) {
+	if err := c.Kill(ctx, syscall.SIGKILL); err != nil {
+		t.Helper()
+		t.Fatalf("could not kill container %q: %v", c.ID(), err)
+	}
+}
+
+func deleteContainer(ctx context.Context, t testing.TB, c *hcsv2.Container) {
+	if err := c.Delete(ctx); err != nil {
+		t.Helper()
+		t.Fatalf("could not delete container %q: %v", c.ID(), err)
+	}
+}
+
+func cleanupContainer(ctx context.Context, t testing.TB, host *hcsv2.Host, c *hcsv2.Container) {
+	deleteContainer(ctx, t, c)
+	removeContainer(ctx, t, host, c.ID())
+}
+
+//
+// runtime
+//
+
+func listContainerStates(_ context.Context, t testing.TB, rt runtime.Runtime) []runtime.ContainerState {
+	css, err := rt.ListContainerStates()
+	if err != nil {
+		t.Helper()
+		t.Fatalf("could not list containers: %v", err)
+	}
+
+	return css
+}
+
+// assertNumberContainers asserts that n containers are found, and then returns the container states.
+func assertNumberContainers(ctx context.Context, t testing.TB, rt runtime.Runtime, n int) {
+	fmt := "found %d running containers, wanted %d"
+	css := listContainerStates(ctx, t, rt)
+	nn := len(css)
+	if nn != n {
+		t.Helper()
+
+		if nn == 0 {
+			t.Fatalf(fmt, nn, n)
+		}
+
+		cs := make([]string, nn)
+		for i, c := range css {
+			cs[i] = c.ID
+		}
+
+		t.Fatalf(fmt+":\n%#+v", nn, n, cs)
+	}
+}
+
+func getContainerState(ctx context.Context, t testing.TB, rt runtime.Runtime, id string) runtime.ContainerState {
+	css := listContainerStates(ctx, t, rt)
+
+	for _, cs := range css {
+		if cs.ID == id {
+			return cs
+		}
+	}
+
+	t.Helper()
+	t.Fatalf("could not find container %q", id)
+	return runtime.ContainerState{} // just to make the linter happy
+}
+
+func assertContainerState(ctx context.Context, t testing.TB, rt runtime.Runtime, id, state string) {
+	cs := getContainerState(ctx, t, rt, id)
+	if cs.Status != state {
+		t.Helper()
+		t.Fatalf("got container %q status %q, wanted %q", id, cs.Status, state)
+	}
+}
+
+//
+// mount management
+//
+
+func mountRootfs(ctx context.Context, t testing.TB, host *hcsv2.Host, id string) (scratch string, rootfs string) {
+	scratch = filepath.Join(guestpath.LCOWRootPrefixInUVM, id)
```

Comment:
Is there a way we could refactor the normal code flow to somehow share this function?

Reply:
It should be, yeah
Itll create another PR that tries to leverage normal code flow as much as possible